src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStack.cs

```diff
@@ -14,51 +14,179 @@ internal struct ReadStack
+        // The bag of preservable references.
         public DefaultReferenceResolver ReferenceResolver;
 
-        public void Push()
+        /// <summary>
+        /// Whether we need to read ahead in the inner read loop.
+        /// </summary>
+        public bool SupportContinuation;
+
+        private void AddCurrent()
         {
             if (_previous == null)
             {
                 _previous = new List<ReadStackFrame>();
             }
 
-            if (_index == _previous.Count)
+            if (_count > _previous.Count)
             {
                 // Need to allocate a new array element.
                 _previous.Add(Current);
             }
             else
             {
-                Debug.Assert(_index < _previous.Count);
-
                 // Use a previously allocated slot.
-                _previous[_index] = Current;
+                _previous[_count - 1] = Current;
             }
 
-            Current.Reset();
-            _index++;
+            _count++;
         }
 
-        public void Pop()
+        public void InitializeRoot(Type type, JsonSerializerOptions options)
         {
-            Debug.Assert(_index > 0);
-            Current = _previous[--_index];
+            JsonClassInfo jsonClassInfo = options.GetOrAddClass(type);
+            Debug.Assert(jsonClassInfo.ClassType != ClassType.Invalid);
+
+            Current.JsonClassInfo = jsonClassInfo;
+
+            // The initial JsonPropertyInfo will be used to obtain the converter.
+            Current.JsonPropertyInfo = jsonClassInfo.PolicyProperty!;
+
+            if (options.ReferenceHandling.ShouldReadPreservedReferences())
+            {
+                ReferenceResolver = new DefaultReferenceResolver(writing: false);
+            }
         }
 
-        public bool IsLastFrame => _index == 0;
+        public void Push()
+        {
+            if (_continuationCount == 0)
+            {
+                if (_count == 0)
+                {
+                    // The first stack frame is held in Current.
+                    _count = 1;
+                }
+                else
+                {
+                    JsonClassInfo jsonClassInfo;
+                    if ((Current.JsonClassInfo.ClassType & (ClassType.Object | ClassType.Value | ClassType.NewValue)) != 0)
+                    {
+                        // Although ClassType.Value doesn't push, a custom custom converter may re-enter serialization.
+                        jsonClassInfo = Current.JsonPropertyInfo!.RuntimeClassInfo;
+                    }
+                    else
+                    {
+                        jsonClassInfo = Current.JsonClassInfo.ElementClassInfo!;
+                    }
+
+                    AddCurrent();
+                    Current.Reset();
+
+                    Current.JsonClassInfo = jsonClassInfo;
+                    Current.JsonPropertyInfo = jsonClassInfo.PolicyProperty!;
```

Comment:
Similar question. What's ensuring this is not null.

Reply:
A non-null classInfo.PolicyProperty is a common pattern throughout the code including pre-existing in master. This PR didn't change that.